lib/pacer_test.go

```diff
@@ -0,0 +1,224 @@
+//        |    ,'xxxxxx|      `.
+//        |  ,'xxxxxxxx|        `.
+//        | /xxxxxxxxxx|          \
+//        |/xxxxxxxxxxx|           \
+//  Mean -+-------------------------\--------------------------> t
+//        |                          \           |xxxxxxxxxxx/
+//        |                           \          |xxxxxxxxxx/
+//        |                            `.        |xxxxxxxx,'
+//        |                              `.      |xxxxxx,'
+//  Mean  |                                `-.   |xxx,-'
+//  -Amp -|                                   `-,,,-'
+//
+// The four multipliers are how many multiples of ampHits() away from
+// Mean*t the integral is after 1, 2, 3 and 4 quarter-periods respectively.
+var qpahm = map[float64][]float64{
+	MeanUp:   {1, 2, 1, 0},
+	Peak:     {1, 0, -1, 0},
+	MeanDown: {-1, -2, -1, 0},
+	Trough:   {-1, 0, 1, 0},
+}
+
+// Helper struct type to make creating SinePacers easier
+type sineTest struct {
+	period int // Period, in seconds
+	mean   int // Mean request rate, in hits/sec
+	amp    int // Amplitude, in hits/sec
+}
+
+func (st sineTest) Pacer(startAt float64) SinePacer {
+	return SinePacer{
+		Period:  time.Duration(st.period) * time.Second,
+		Mean:    ConstantPacer{st.mean, time.Second},
+		Amp:     ConstantPacer{st.amp, time.Second},
+		StartAt: startAt,
+	}
+}
+
+// See comment for qpahm above for why this is useful.
+func (st sineTest) ampHits() float64 {
+	return float64(st.amp) * float64(st.period) / (2 * math.Pi)
+}
+
+func TestSinePacerHits(t *testing.T) {
+	tests := []sineTest{
+		// {period in secs, mean hits/sec, amp hits/sec}
+		{20 * 60, 100, 90},
+		{60, 1000, 10},
+		{1, 10, 7},
+		{1, 1, 0},
+		{1e6, 10, 7},
+		{60, 1000, 999},
+	}
+
+	for ti, tt := range tests {
+		for name, startAt := range quarterPeriods {
+			sp := tt.Pacer(startAt)
+			// See comment for qpahm (quarter-period ampHits multiplier) above.
+			for i, mult := range qpahm[startAt] {
+				periods := i + 1
+				want := float64(tt.mean*periods*tt.period)/4 + tt.ampHits()*mult
+				if got := sp.hits(time.Duration(periods) * sp.Period / 4); !floatEqual(got, want) {
+					t.Errorf("%d(%s): %+v.hits(%d/4 period) = %g, want %g",
+						ti, name, sp, i+1, got, want)
+				}
+			}
+		}
+	}
+
+	// TestSinePacerInvalid takes care of most of the sad path.
+	sp := sineTest{1, 1, 0}.Pacer(0)
+	if got := sp.hits(-1); got != 0 {
+		t.Errorf("%d: %+v.hits(-1) = %g, want 0", len(tests), sp, got)
+	}
+}
+
+func TestSinePacerInvalid(t *testing.T) {
+	tests := []sineTest{
+		// {period in secs, mean hits/sec, amp hits/sec}
+		{0, 100, 90},   // Zero period
+		{60, 0, 90},    // Zero mean
+		{60, 100, 110}, // Amp > mean
+		{-10, 100, 90}, // Negative period
+		{60, -10, 90},  // Negative mean
+	}
+
+	for ti, tt := range tests {
+		sp := tt.Pacer(0)
+		if got := sp.hits(sp.Period); got != 0 {
+			t.Errorf("%d: %+v.hits(%s) = %g, want 0",
+				ti, sp, sp.Period, got)
+		}
+	}
+}
+
+func TestSinePacerPace_Flat(t *testing.T) {
```

Comment:
What is this test covering exactly?

Reply:
It validates that the code does the right thing when there's no amplitude, i.e. the sine wave is actually a straight line. I've added a comment indicating this.